Add EMNIST Dataset for Image Classification (issue #129)

DerrickUnleashed opened the pull request:
## This PR adds support for the EMNIST dataset’s Balanced split, extending the existing MNIST datasets.

- Uses the existing download and caching infrastructure for consistency and efficiency.  
- Supports familiar parameters aligned with MNIST datasets.  
- Includes comprehensive tests modeled after MNIST test suites.  
- Updates documentation with examples and references.

cregouby commented on R/dataset-mnist.R:
```diff
@@ -149,6 +149,160 @@ kmnist_dataset <- dataset(
+#' @examples
+#' \dontrun{
+#' # Create the EMNIST Balanced training dataset, downloading if needed
+#' ds <- emnist_dataset(root = "~/data", train = TRUE, download = TRUE)
```
**todo**: we should avoid `~/data` path, because it is not system agnostic.  
**question** : I would't configure a `root` in the minimal example, as the default cache folder should be good enough.

DerrickUnleashed commented:
Thanks a lot for the warm welcome!   
I've made all the requested changes:

-  Updated the parameter to use `split = c(...)` instead of `train = TRUE`.
-  Ran `devtools::document()` to update the `NAMESPACE` and man files.
-  Added an entry in `NEWS.md` for the EMNIST dataset.
-  Added myself to the `Authors@R` with `role = c("ctb")`.

Let me know if anything else needs tweaking!

cregouby commented on R/dataset-mnist.R:
```diff
@@ -149,6 +149,178 @@ kmnist_dataset <- dataset(
+#'
+#' @examples
+#' \dontrun{
+#' ds <- emnist_dataset(root = "~/data", split = "balanced", download = TRUE)
```
**todo**: we should avoid ~/data path, because it is not system agnostic.
question : I would't configure a root in the minimal example, as the default cache folder should be good enough.

DerrickUnleashed replied:
Avoided using `~/data` path as it's not system-agnostic and can cause issues across environments.

cregouby commented on .RData:
**todo**: please do not commit such file

DerrickUnleashed replied:
Acknowledged and removed the file from the commit. Will ensure only relevant files are included.

cregouby commented on man/transform_perspective.Rd:
**todo**: please do not commit file unrelated to the P.R. content

DerrickUnleashed replied:
Acknowledged and removed the file from the commit. Will ensure only relevant files are included.

cregouby commented on man/transform_random_perspective.Rd:
**todo**: please do not commit file unrelated to the P.R. content

DerrickUnleashed replied:
Acknowledged and removed the file from the commit. Will ensure only relevant files are included.

cregouby commented on man/transform_random_resized_crop.Rd:
**todo**: please do not commit file unrelated to the P.R. content

DerrickUnleashed replied:
Acknowledged and removed the file from the commit. Will ensure only relevant files are included.

cregouby commented on man/transform_resize.Rd:
**todo**: please do not commit file unrelated to the P.R. content

cregouby commented on man/transform_resized_crop.Rd:
**todo**: please do not commit file unrelated to the P.R. content

cregouby commented:
Thanks !
Can you please resolve the merge conflicts with `main` ?

cregouby commented on R/dataset-mnist.R:
```diff
@@ -149,6 +149,175 @@ kmnist_dataset <- dataset(
+#'
+#' @examples
+#' \dontrun{
+#' ds <- emnist_dataset(split = "balanced", download = TRUE)
+#' sample <- ds[1]
+#' image <- sample[[1]]
+#' label <- sample[[2]]
```
**suggestion** : As a first time user, I want to **see** what is in this dataset item 1, right now the affectation in not showing anything but only providing a taste of a better name to the object. I think we can do both. Plus we can highlight that the two values in the list are named. What about
```suggestion
#' emnist <- emnist_dataset(split = "balanced", download = TRUE)
#' first_item <- emnist[1]
#' # image in item 1
#' first_item$x
#' # label of item 1
#' first_item$y
```

DerrickUnleashed commented:
I've made the necessary changes to address the issue with `emnist_dataset()`, including adding a default `root` and a user message to inform about the download.

However, while testing, I noticed similar issues in other dataset examples as well. Would you like me to go ahead and fix those too?

```r
Error in mnist_dataset(download = TRUE) : 
  argument "root" is missing, with no default
```
```r
Error in kmnist_dataset(download = TRUE) : 
  argument "root" is missing, with no default
```
```r
Error in fashion_mnist_dataset(download = TRUE) : 
  argument "root" is missing, with no default
```

cregouby commented:
> I've made the necessary changes to address the issue with `emnist_dataset()`, including adding a default `root` and a user message to inform about the download.
> 
> However, while testing, I noticed similar issues in other dataset examples as well. Would you like me to go ahead and fix those too?
> 
> ```r
> Error in mnist_dataset(download = TRUE) : 
>   argument "root" is missing, with no default
> ```
> 
> ```r
> Error in kmnist_dataset(download = TRUE) : 
>   argument "root" is missing, with no default
> ```
> 
> ```r
> Error in fashion_mnist_dataset(download = TRUE) : 
>   argument "root" is missing, with no default
> ```

Yes, of course.

cregouby commented:
fix #129 

could you please update the documentation with devtool::document() ?

DerrickUnleashed commented:
Thanks for the fix! I've updated the documentation using `devtools::document()`. Let me know if there's anything else needed.

DerrickUnleashed commented:
Thanks! That makes sense. Since these issues with `mnist_dataset()`, `kmnist_dataset()`, and `fashion_mnist_dataset()` are slightly different from the one addressed in the current PR, would you prefer we create a new issue for them? Or should I go ahead and include the fixes in this PR itself? Let me know what works best.
> > I've made the necessary changes to address the issue with `emnist_dataset()`, including adding a default `root` and a user message to inform about the download.
> > However, while testing, I noticed similar issues in other dataset examples as well. Would you like me to go ahead and fix those too?
> > ```r
> > Error in mnist_dataset(download = TRUE) : 
> >   argument "root" is missing, with no default
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > ```r
> > Error in kmnist_dataset(download = TRUE) : 
> >   argument "root" is missing, with no default
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > ```r
> > Error in fashion_mnist_dataset(download = TRUE) : 
> >   argument "root" is missing, with no default
> > ```
> 
> Yes, of course.

DerrickUnleashed commented:
**Suggestion : ** Should we consider adding `.RData` to the `.gitignore` file?

DerrickUnleashed commented on R/dataset-mnist.R:
```diff
@@ -218,3 +196,203 @@ fashion_mnist_dataset <- dataset(
+    mnist = as.character(0:9)
+  ),
+
+  initialize = function(root = rappdirs::user_cache_dir("torch"), split = "balanced", transform = NULL, target_transform = NULL,
```
Thank you. I've updated the design to use ```tempdir()``` for temporary download folders, preventing the mixing of temporary and cached files and allowing for intelligent cache cleaning.

cregouby reviewed (commented):
**praise** : Thanks for the addition. We are almost there, we just need to lower the test time and footprint.

cregouby commented:
Sorry, but the `skip_on_cran()` is missing and your last commit despite the improvement removes the code coverage on 5 different values of the "split". We need those tests to be covered. Actually, whatever the doanwload= value, the archive is downloaded only once within the test making void any optimization (like to switch to download = FALSE). Could you please fix that ?

cregouby reviewed (changes requested):
**todo** Sorry, but the `skip_on_cran()` is missing
**todo** your last commit despite the improvement, removes the code coverage on 5 different values of the "split =". We need those tests to be covered. Actually, whatever the download= value, the archive is downloaded only once within the test, making void any optimization (like to switch to download = FALSE). Could you please fix that ?

DerrickUnleashed commented:
Thanks for pointing that out. I've just added back the `skip_on_cran()` (it was removed by mistake in my previous commit). I've also updated the tests to restore full coverage for the different `split = values`. The tests now ensure that the archive is only downloaded once. Let me know if anything else needs adjustment.

DerrickUnleashed commented on tests/testthat/test-dataset-mnist.R:
```diff
@@ -77,6 +77,47 @@ test_that("fashion_mnist_dataset loads correctly", {
+  for (split in names(splits)) {
+    info <- splits[[split]]
+
+    ds <- emnist_dataset(dir, split = split, download = TRUE)
```
I understood my error I have forgotten to use the `skip()` command after using `skip_on_cran()` now it has less test footprint. (has reduced from ~15 mins to ~9 mins)

DerrickUnleashed commented:
I understood my error I have forgotten to use the `skip()` command after using `skip_on_cran()` now it has less test footprint. (has reduced from ~15 mins to ~9 mins)

DerrickUnleashed commented:
@cregouby I have made the requested changes...Is there anything else that needs to be updated ?

DerrickUnleashed commented on R/dataset-mnist.R:
```diff
@@ -382,3 +382,206 @@ fashion_mnist_dataset <- dataset(
+    rlang::inform(glue::glue(
+      "Preparing to download EMNIST dataset.\n",
+      "- Size: ~0.5GB compressed\n",
+      "- Expected downloading time: ~6 minutes\n",
```
Thanks for the suggestion... We'll avoid mentioning download time since it's not reliably measurable.
